### PR TITLE
FEATURE: Centralized 2FA page

### PR DESCRIPTION
**Pull request by OsamaSayegh**
This PR introduces a centralized 2FA page that can be used for any actions that require second factor
authentication before a user is allowed to perform them. Such actions include
granting another user admin access, changing password and signing in. In a more
technical sense, an action is the logic encapsulated in a Rails controller
action without the logic related to 2fa enforcement/handling.

The bulk of this logic is handled in SecondFactor::AuthManager which
has extensive comments on how all the different flows are handled. This
PR in particular only changes the "grant admin" action to use the new centralized
2FA page, the login flows are not changed.

### CODE REVIEW

File: app/controllers/admin/users_controller.rb [hunk 1]
```diff
@@ -191,25 +190,31 @@ def revoke_admin
...
-        if current_user.security_keys_enabled?
-          Webauthn.stage_challenge(current_user, secure_session)
-          failure_payload.merge!(Webauthn.allowed_credentials(current_user, secure_session))
-        end
-        render json: failed_json.merge(failure_payload)
-      end
-    else
-      AdminConfirmation.new(@user, current_user).create_confirmation
+    manager = SecondFactorVerificationManager.new(current_user)
+    manager.on_no_second_factors_enabled do
```
Comment (SamSaffron) [inline app/controllers/admin/users_controller.rb:194]: this block doing a render is a bit odd, can we avoid it being responsible for render?

File: app/controllers/admin/users_controller.rb [hunk 2]
```diff
@@ -191,25 +190,31 @@ def revoke_admin
...
     end
+
+    manager.on_second_factor_auth_required do |config|
+      user = User.find_by(id: params[:user_id])
+      guardian.ensure_can_grant_admin!(user)
+      config.callback_params = { user_id: user.id }
+      config.redirect_path = admin_user_show_path(id: user.id, username: user.username)
+    end
+
+    manager.on_second_factor_auth_successful do |callback_params|
```
Comment (SamSaffron) [inline app/controllers/admin/users_controller.rb:208]: Same... another block rendering, can the API avoid that somehow?

File: app/controllers/session_controller.rb [hunk 3]
```diff
@@ -424,6 +426,72 @@ def one_time_password
     render layout: 'no_ember', locals: { hide_auth_buttons: true }
   end
 
+  def second_factor_auth_show
+    user = current_user
```
Comment (martin-brennan) [inline app/controllers/session_controller.rb:430]: You can just use `requires_login` for this, no?

https://github.com/discourse/discourse/blob/b5765ad95b5fa80a09cda4325ed5c3776ddf4440/app/controllers/tags_controller.rb#L14-L21

**Comment by pmusaraj** [comment]
@OsamaSayegh can you post some screenshots of what this page looks like (in different states and with zero, one or more 2FA mechanisms enabled)?

**Comment by OsamaSayegh** [comment]
With one 2fa method enabled (security keys):

image


With 2 2fa methods enabled (security key + TOTP):

image

Clicking the "Use an authenticator app instead" swaps the security key form with the TOTP form:

image

When 2fa challenge is completed successfully:

image

And when it fails (bad token):

image

With all 3 2fa methods enabled (security key + TOTP + backup codes):

image

image

When there are zero 2fa methods enabled, there is no way to access the page so no screenshots :)

**Review by martin-brennan — CHANGES_REQUESTED** [review]
IMO we should have a nice description/commit message for this PR explaining why we are centralizing, as well as a brief overview of how the new auth flow and `AuthManager` code works. This will provide important context for when a developer comes along later to `git blame` 

Also this doesn't need to be done in this PR, but I wonder if we should say why we are requiring 2FA for a certain thing...e.g. when I went to grant admin I got this:

image

Which is fine, but maybe we should have a blurb like "Re-authentication is required to **grant admin**." with the text in bold the action being taken.

Overall I think this is great just a few things to fix up, my main concern is the route made for testing only

File: app/controllers/session_controller.rb [hunk 4]
```diff
@@ -424,6 +428,128 @@ def one_time_password
...
+    end
+    render json: {
+      ok: true,
+      callback_method: challenge[:callback_method],
+      callback_path: challenge[:callback_path],
+      redirect_path: challenge[:redirect_path]
+    }, status: 200
+  end
+
+  if Rails.env.test?
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply app/controllers/session_controller.rb:538]: This route is meant to be an auxiliary route for the other 2 new routes in session controller, `second_factor_auth_show` and `second_factor_auth_perform`; it's not meant to replace tests for any real routes that require 2fa (if you do ctrl+f for `/session/2fa/test-action` in this PR you'll only find it used in `session_controller_spec.rb`).

For example, I need to write test case for when the `SessionController#second_factor_auth_perform` action receives a bad TOTP. In order to do this I need something that sets up a challenge in the session before I can actually send a request the `second_factor_auth_perform` action with a bad TOTP because without a challenge the `second_factor_auth_perform` would respond with 404 right away. This is where the auxiliary route comes in handy:

1. first I send a request to the auxiliary route and it takes care of setting up a challenge
2. now with the challenge in place I can a send a request to the `second_factor_auth_perform` with a bad TOTP
3. and then assert that the `second_factor_auth_perform` action responds with the right status code and error message.

See this particular flow that I've described [here](https://github.com/discourse/discourse/pull/15377/files#diff-62424ac8e3831860baf82b385e194763d10d28a036b5f0f805e947b81f4728e4R2426) and there are a few more flows that test specific edge cases in the `second_factor_auth_perform` action. Without this route, I'd need to call some other action from another controller (currently the only option would be grant admin which requires current user to b admin) in the tests of the session controller and I think that'd mix things up and cause confusion as to what the test is testing.

> we are not testing the thing that will actually happen

There is a bunch of tests in `users_controller_spec.rb` that test the whole 2fa flow for the grant admin route and the plan is to add tests like these for every route that we migrate to this new 2fa system.

File: app/controllers/session_controller.rb [hunk 5]
```diff
@@ -674,4 +800,31 @@ def redeem_invitation(invite, sso)
     Rails.logger.warn("SSO invite redemption failed: #{e}")
     raise Invite::RedemptionFailed
   end
+
+  def find_second_factor_challenge(nonce)
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply app/controllers/session_controller.rb:804]: Will do. 

> checks if the challenge was successful, not sure if that should also be here

No it shouldn't do that here because the challenge at this stage is not successful, but will be marked successful if the 2fa succeeds.

File: lib/second_factor/auth_manager_result.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+class SecondFactor::AuthManagerResult
+  class StrictEnum
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply lib/second_factor/auth_manager_result.rb:4]: When I initially added this class, I needed it to do that but then I changed a lot of things in the result class that we ended up not needing the whole strict enum thing. Will remove.

File: app/controllers/session_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (martin-brennan) [reply app/controllers/session_controller.rb:538]: Ah thank you sorry I must have missed the ones in `users_controller_spec.rb`. Make sense now thanks for taking the time to explain :+1:

**Comment by OsamaSayegh** [comment]
Ok @martin-brennan, can you take another look please? thanks!

**Review by martin-brennan — APPROVED** [review]
Looks great, thanks for making those changes, absolute beast of a PR! I just added a quick PR description so it can be used when we squash into one commit. Nice one :+1:

File: spec/lib/second_factor/actions/grant_admin_spec.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe SecondFactor::Actions::GrantAdmin do
+  fab!(:admin) { Fabricate(:admin) }
+  fab!(:user) { Fabricate(:user) }
+
+  before do
+    AdminConfirmation.cleanup_redis
```
Comment (tgxworld) [inline spec/lib/second_factor/actions/grant_admin_spec.rb:10]: I think we always want to be cleaning up after and not before. If you need to clean up before a test, it means that we're already leaking state across tests which can make our test flaky.

File: spec/requests/session_controller_spec.rb [hunk 8]
```diff
@@ -2260,4 +2260,195 @@ def post_login
...
+      sign_in(user)
+    end
+
+    it 'returns 404 if there is no challenge for the given nonce' do
+      get "/session/2fa.json", params: { nonce: 'asdasdsadsad' }
+      expect(response.status).to eq(404)
+      expect(response.parsed_body["error"]).to eq(I18n.t("second_factor_auth.challenge_not_found"))
+    end
+
+    it 'returns 404 if the nonce does not match the challenge nonce' do
```
Comment (tgxworld) [inline spec/requests/session_controller_spec.rb:2277]: Should we split this test up into 2? There are two assertions in this test but the description of the test is only describing one.

File: spec/requests/session_controller_spec.rb [hunk 9]
```diff
@@ -2260,4 +2260,195 @@ def post_login
...
+      post "/session/2fa/test-action"
+      expect(response.status).to eq(403)
+      expect(response.parsed_body["second_factor_challenge_nonce"]).to be_present
```
Comment (tgxworld) [inline spec/requests/session_controller_spec.rb:2280]: Just curious, what is the purpose of this request in relation to the request below? Should we perhaps split this test up into two?

File: spec/requests/session_controller_spec.rb [hunk 10]
```diff
@@ -2260,4 +2260,195 @@ def post_login
...
+      get "/session/2fa.json", params: { nonce: 'wrongnonce' }
+      expect(response.status).to eq(404)
+      expect(response.parsed_body["error"]).to eq(I18n.t("second_factor_auth.challenge_not_found"))
+    end
+
+    it 'returns 401 if the challenge has expired' do
```
Comment (tgxworld) [inline spec/requests/session_controller_spec.rb:2286]:
```suggestion
    it 'returns 401 if the challenge nonce has expired' do
```

File: lib/second_factor/auth_manager.rb [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+  end
+
+  def initialize(current_user, guardian, action)
+    @current_user = current_user
+    @guardian = guardian
+    @action = action
+    @allowed_methods = Set.new([
+      UserSecondFactor.methods[:totp],
+      UserSecondFactor.methods[:security_key],
+    ]).freeze
```
Comment (tgxworld) [inline lib/second_factor/auth_manager.rb:136]: Just curious, what is the purpose of `freeze` here?

File: spec/lib/second_factor/actions/grant_admin_spec.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  def params(hash)
+    ActionController::Parameters.new(hash)
+  end
+
+  def create_instance(u)
+    SecondFactor::Actions::GrantAdmin.new(u, Guardian.new(u))
+  end
```
Comment (tgxworld) [inline spec/lib/second_factor/actions/grant_admin_spec.rb:19]:
```suggestion
  def create_instance(user)
    SecondFactor::Actions::GrantAdmin.new(user, Guardian.new(user))
  end
```

File: lib/second_factor/actions/base.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module SecondFactor::Actions
+  class Base
+    include Rails.application.routes.url_helpers
+    attr_reader :current_user, :guardian
+
+    def initialize(current_user, guardian)
```
Comment (tgxworld) [inline lib/second_factor/actions/base.rb:8]: Are we able to just pass in the `guardian` instead since the guardian will always be created from the current user?

File: lib/second_factor/auth_manager.rb [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    if generated_at < MAX_CHALLENGE_AGE.ago.to_i
+      raise SecondFactor::BadChallenge.new(
+        "second_factor_auth.challenge_expired",
+        status_code: 401
+      )
+    end
+    challenge
+  end
+
+  def initialize(current_user, guardian, action)
```
Comment (tgxworld) [inline lib/second_factor/auth_manager.rb:129]: I have the same concerns here, will we ever pass in a guardian that is not created from the current user?

File: lib/second_factor/auth_manager.rb [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+      )
+    end
+
+    secure_session["current_second_factor_auth_challenge"] = nil
+    callback_params = challenge[:callback_params]
+    @action.second_factor_auth_completed!(callback_params)
+  end
+
+  def add_method(id)
+    if !@allowed_methods.include?(id)
```
Comment (tgxworld) [inline lib/second_factor/auth_manager.rb:191]: Not sure if this is necessary since we're already using a set.

File: spec/lib/second_factor/auth_manager_spec.rb [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+      nonce, secure_session = stage_challenge(successful: true)
+
+      action = create_action
+      manager = create_manager(action)
+      request = create_request(
+        request_method: "POST",
+        path: "/abc/xyz"
+      )
+
+      freeze_time 6.minutes.from_now
```
Comment (tgxworld) [inline spec/lib/second_factor/auth_manager_spec.rb:171]: I think it'll be better to use the `SecondFactor::AuthManager::MAX_CHALLENGE_AGE` constant here so that these tests don't randomly break if we were to tweak the constant.

File: spec/requests/session_controller_spec.rb (hunk 9 above)
Comment (OsamaSayegh) [reply spec/requests/session_controller_spec.rb]: This test case is for the `#second_factor_auth_show` action which requires a 2fa challenge to exist in the session, otherwise it'd just respond with a 404. To be able to test that action, I need something to set up a challenge in the session for me, and that's what the post request to `/session/2fa/test-action` does. The assertions for the response/status code of that helper/test action are not really useful, so I've removed them.

File: lib/second_factor/auth_manager.rb (hunk 15 above)
Comment (OsamaSayegh) [reply lib/second_factor/auth_manager.rb:191]: The set is frozen (see why above) so yes we do need this check unless you think freezing the set doesn't add much value?

File: spec/rails_helper.rb [hunk 17]
```diff
@@ -510,3 +511,32 @@ class << self
     attr_accessor :value
   end
 end
+
+class TestSecondFactorAction < SecondFactor::Actions::Base
+  def self.called_methods
```
Comment (eviltrout) [inline spec/rails_helper.rb:516]: This is a bit of an odd pattern to me. It seems like you are re-implementing stubs yourself. Rather than doing this, could we not use `expects` and `stubs` in your tests?

File: lib/admin_confirmation.rb [hunk 18]
```diff
@@ -59,4 +59,10 @@ def self.find_by_code(token)
     ac
   end
 
+  def self.cleanup_redis
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply lib/admin_confirmation.rb]: I moved this method to `grant_admin_spec.rb`.

File: config/routes.rb [hunk 19]
```diff
@@ -371,6 +371,11 @@ def patch(*) end # Disable PATCH requests
     post "session/email-login/:token" => "session#email_login"
     get "session/otp/:token" => "session#one_time_password", constraints: { token: /[0-9a-f]+/ }
     post "session/otp/:token" => "session#one_time_password", constraints: { token: /[0-9a-f]+/ }
+    get "session/2fa" => "session#second_factor_auth_show"
+    post "session/2fa" => "session#second_factor_auth_perform"
+    if Rails.env.test?
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply config/routes.rb:376]: If we do it in a constraint, then Rails would be "aware" of the route and the router could check incoming requests against this route in production, which wouldn't be the end of the world but I think it's a tiny bit less than ideal. Guarding the route with an `if` condition makes Rails completely unaware of the route in production.

File: app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+import pretender from "discourse/tests/helpers/create-pretender";
+import { SECOND_FACTOR_METHODS } from "discourse/models/user";
+import { test } from "qunit";
+
+const { TOTP, BACKUP_CODE, SECURITY_KEY } = SECOND_FACTOR_METHODS;
+
+function setupResponse(response = {}) {
+  const status = response.status || 200;
+  delete response.status;
+  pretender.get("/session/2fa.json", () => {
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js:16]: I thought this would be a very simple change, but it turned that I needed to restructure this test file. I'm happy to do that, but I just want to double check this first before I continue:

So there are a lots of different states that the `/session/2fa/` page can be in depending on the available 2fa methods for the current user. The page loads the 2fa details of the user from `/session/2fa.json` and based on the server response it determines what to show and what to hide etc. which that means I need pretender to return different responses for the same route to be able test the various states of the page.

Since `needs.pretender` works at the module (I think each `acceptance` block is called a "module"?) level, I'd have to create a module for every different state. For example this test case would need to be split into 3 modules (one for each `setupResponse` call):

https://github.com/discourse/discourse/blob/ecc82cbe4fb55889db285efac888996122996294/app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js#L36-L98

I see most of our acceptance test files have 1 or 2 modules which makes me a little bit unsure if I should create 10+ modules in a single file?

Also just to clarify what you mean by proper clean up: if I use pretender directly like I did here, does it leak my responses to other tests?

File: config/routes.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (eviltrout) [reply config/routes.rb:376]: Cool I think your justificaiton here is good.

File: app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (eviltrout) [reply app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js:16]: I'm not sure if pretender leaks currently doing it this way, but I can't guarantee it won't in the future. It's an API we don't want to support.

> which that means I need pretender to return different responses for the same route to be able test the various states of the page.

Did you know pretender can return things conditionally, depending on the input? We commonly use it to reply with fixtures but it's quite powerful. Here's a very simple example:

https://github.com/discourse/discourse/blob/main/app/assets/javascripts/discourse/tests/helpers/create-pretender.js#L146

I think this could help you very much in this situation!

[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js:16]: > Did you know pretender can return things conditionally, depending on the input?

Yes, I'm aware of that, but the `/session/2fa.json` endpoint takes no parameters; it just responds with the 2fa details of the current user. But from user to user, these details can differ a lot and therefore the 2fa page will look different as well.

I guess I could add a tests-only param that the pretender can check and return the right response for the param value? Something like this:

```javascript
await visit("/session/2fa?pretender_response=response11");
```

Then in the pretender:

```javascript
pretender.get("/session/2fa.json", (request) => {
  return RESPONSES[request.params.pretender_response];
});
```

Or maybe we keep a global variable in the test file that indicates which response the pretender should give, and then each test case changes that global variable to the id of the response it needs?

I like the global variable approach because it can be done without changing anything in non-test files. What's your preference?

[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply app/assets/javascripts/discourse/tests/acceptance/second-factor-auth-test.js]: Ok I've made this change (the one based on the nonce param). @eviltrout can you please take another look (on the pretender bit and the PR in general) and let me know if you're happy with it? thanks!

**Review by eviltrout — APPROVED** [review]
Awesome, thank you.

**Comment by OsamaSayegh** [comment]
Thank you for the reviews everyone! A ton of improvements came out of this! Merging the PR now 🚀
